src/remount-fs/remount-fs.c

```diff
@@ -69,11 +70,87 @@ static int do_remount(const char *path, bool force_rw, Hashmap **pids) {
+
+        const char *uuid;
+        r = sd_device_get_property_value(device, "ID_FS_UUID", &uuid);
+        if (r < 0)
+                return log_warning_errno(r, "Failed to get filesystem UUID for device %s: %m", path);
+
+        r = sd_id128_from_string(uuid, ret_uuid);
+        if (r < 0)
+                return log_warning_errno(r, "Failed to parse ID_FS_UUID '%s' for device %s: %m",
+                                         uuid, path);
+        return 0;
+}
+
+static int mount_point_matches(const struct mntent *me) {
+        _cleanup_free_ char *what = NULL, *where = NULL, *type = NULL;
+        int r;
+
+        r = path_get_mount_info(ASSERT_PTR(me)->mnt_dir, &what, &where, &type, /* options= */ NULL);
+        if (r < 0)
+                return log_warning_errno(r, "Failed to acquire information about mount point %s: %m", me->mnt_dir);
+        if (!path_equal(me->mnt_dir, where)) { /* We got the containing mount point */
+                log_info("Mount point \"%s\" is not mounted.", me->mnt_dir);
+                return false;
+        }
+
+        if (!streq(type, me->mnt_type) && !streq(me->mnt_type, "auto")) {
+                log_info("File system mounted at \"%s\" doesn't match fstab (type %s vs. %s), ignoring.",
+                         me->mnt_dir, type, me->mnt_type);
+                return false;
+        }
+
+        if (!fstype_is_blockdev_backed(type)) {
+                log_debug("File system mounted at \"%s\" matches fstab.", me->mnt_dir);
+                return true; /* Nothing more to check, the "device" is not important. */
+        }
+
+        /* Figure out if the two paths point at the same device node */
+        _cleanup_free_ char *node = fstab_node_to_udev_node(me->mnt_fsname);
+        if (!node)
+                return log_oom();
+
+        r = path_equal_or_inode_same_full(what, node, AT_NO_AUTOMOUNT);
+        if (r < 0)
+                return log_warning_errno(r, "Failed to determine if %s and %s for mount point \"%s\" are same: %m",
+                                         what, node, me->mnt_dir);
+        if (r == 0) {
+                /* Try to figure out if the file system identifier matches…
+                 * This way we cover multi-device file systems like btrfs or bcachefs. */
+                sd_id128_t uuid_a, uuid_b;
+
+                r = device_node_uuid(what, &uuid_a);
+                if (r < 0)
+                        return r;
+
+                r = device_node_uuid(node, &uuid_b);
+                if (r < 0)
+                        return r;
+
+                r = sd_id128_equal(uuid_a, uuid_b);
+        }
+
+        if (r > 0)
+                log_debug("File system mounted at \"%s\" matches fstab.", me->mnt_dir);
+        else
+                log_info("File system mounted at \"%s\" doesn't match fstab (%s vs. %s), ignoring.",
+                         me->mnt_dir, what, me->mnt_fsname);
+        return r;
+}
```

Comment:
is there really value in being so struct here regarding the backing device?

i can see that if the fstype is not matching we skip the fstype-specific mount options. But why insist on the mount source? seems completely unnecessary to me, and likely to trigger all kinds of issues because different file systems report backing block devices differently...

I think if you dumb this down to: "do not apply fstype-specific mount options if fstype doesn't match" then this would be a very clear rule, can be explained to anyone, unlike to cause problems for existing setups, deal with anaconda's/fcos' weird setup nicely and hence generally be a safe thing to do, that is minimal enough to not explode in anyone's faces.

Reply:
Hmm. You are essentially saying that "do not apply the options if they were configured for a different file system" is more complicated than "do not apply some of the options, which may be supported by the target file system, if they were configured for a file system of a different type. Ignore the file system identity". I don't think this is better.

I don't particularly like the idea of trying to filter options. In particular, for `x-*` and `X-*`, it might not even be known if the option is fs-type-specific or not.

Maybe this should be turned into a warning if the fs doesn't match.